feat: Add subnet metric support for netplan

ArtsiomAntropau opened the pull request:
<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [x] I have signed the CLA: https://ubuntu.com/legal/contributors
- [x] I have included a comprehensive commit message using the guide below
- [ ] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [x] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [ ] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [ ] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e docs``.
-->


## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->
```
feat: Add support of subnet metric for netplan

This patch enhances netplan configuration by adding support for subnet metrics. It implements metric handling for DHCP configurations by adding the subnet metric to DHCP overrides with the "route-metric" key. For static routes, it establishes a priority system where the route's metric takes precedence, followed by the subnet's metric, and finally defaulting to 100 if neither is specified. The patch also adds subnet metric support for default gateway routes. These changes provide more flexible network configuration options, allowing for better control over routing priorities in cloud-init generated netplan configurations from network config v1. 
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->


## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

blackboxsw commented on cloudinit/net/netplan.py:
```diff
@@ -149,8 +164,10 @@ def _listify(obj, token=" "):
-                    new_route.update({"metric": route.get("metric", 100)})
+                # Priority for metric: 1. route's metric, 2. subnet's metric, 3. default 100
+                route_metric = route.get("metric")
+                if (route_metric or sn_metric) is not None:
+                    new_route.update({"metric": route_metric or sn_metric or 100})
```
The parenethesis in this conditional do not make sense in all cases. Given that metric value can be 0 you'd have a falsy value for route_metric of 0 which will still leave you silently skipping preserving the requested `metric: 0` on the route.

Same with the new_route.update call below, relying on truthiness will be a pitfall if either route_metric or sn_metric are 0.

Maybe it's better to be explicit:
```suggestion
                if route_metric is not None:
                    new_route["metric"] = route_metric
                elif sn_metric is not None:
                    new_route["metric"] = sn_metric
                else:
                    new_route["metric"] = 100
```

Also I have dropped the default of 100. The only case we try to set metric on the new_route is if either route_metric or sn_metric are not None. If they are not-None we should expect to use the value provided at either scope and we won't be falling back to 100.

blackboxsw reviewed (changes requested):
Thank you for your contribution @ArtsiomAntropau  and improving cloud-init.
While generally we try steering individuals toward use of network v2 configuration instead of augmenting cloud-init's network v1 support, I can see the merit of this feature request for environments where adapting network v1 -> v2 may present a larger hurdle than augmenting existing v1 configuration.

That said, I think the truthy conditionals currently present in this PR need to be more explicit as they don't cope with potential 0 values for metric.

In order to best include this type of contribution into cloud-init we will need some very specific things:
1. unittest coverage of changes in tests/unittests/test_net.py to ensure no regression with `tox -e py3`
2. new unittest element covering and example where metric is provided in dhcp4/dhcp6 subnet and route.
3.  schema changes to add "metric" property as allowed in subnets [cloudinit/config/schemas/schema-network-config-v1.json](https://github.com/canonical/cloud-init/blob/25.1/cloudinit/config/schemas/schema-network-config-v1.json#L451)
4. A test case for network schema support in tests/unittests/config/test_schema.py which would validate the new metric key at subnet level doesn't raise a SchemaValidationError

ArtsiomAntropau commented:
@blackboxsw thanks for your fast review, 

you're absolutely right – my target is to use it with config drive and openstack providers, but I don't think that openstack community have any plans to move to the v2 :( 
but for me, it will solve a big problem in which conflict of gateway metrics (same value) of different interfaces makes instance fully unavailable on ubuntu and debian distro

I'm not seeing problem in any of your proposals, so I applied all your suggestions, updated existing tests (because after https://github.com/canonical/cloud-init/pull/6222#discussion_r2093622459 it started to explicitly provide metric in some cases, I don't think it is a big problem) and added new.

blackboxsw commented on cloudinit/net/netplan.py:
```diff
@@ -149,8 +164,14 @@ def _listify(obj, token=" "):
+                elif sn_metric is not None:
+                    new_route["metric"] = sn_metric
+                else:
+                    new_route["metric"] = 100
```
My mistake for suggesting this fallback to a default of 100 when no metric key is presented. This could cause some significant issues with more complex networks if cloud-init decides to set metric of 100 on routes when no `metric` key is provided. I think we should drop that default and only add metric of 100 if specifically provided. Looking over the previous code, it never would have triggered the default of 100 because `route.get()` was only called when "metric" was present in network-config. I've pushed a supplemental commit which drops this unintentional default when no `metric` key is present and fixes the tests.

blackboxsw reviewed (approved):
Thank you @ArtsiomAntropau. I backed out that default `metric` 100 fallback when no `metric` key were provided to avoid risk of introducing metrics where none were requested. Otherwise looks good to me if this works for you. I also corrected JSON schema definition of metric to be an integer type and added docs and examples to represent this metric. If this looks good to you, we can merge it.

ArtsiomAntropau commented:
@blackboxsw thanks for finalising it :) what are the next steps?

blackboxsw commented:
@ArtsiomAntropau no further steps needed now that CI tests passed here. I've merged it and it'll be included in cloud-init version 25.2 which should released by end of next month. It'll be provided via ubuntu stable release updates back to `Jammy (22.04)`. We will publish content in tip of main this week to Ubuntu Questing (25.10) devel series for testing if that is of any help on your platforms.
